File: src/main/java/dsl/BulkMappingBuilder.java
package dsl;

import java.util.*;
import java.util.function.BiFunction;
import java.util.function.Function;

/**
 * Builder for bulk mapping operations.
 */
public class BulkMappingBuilder {
    private final TargetBuilder parent;
    private final String sourcePattern;
    private final Set<String> exclusions = new HashSet<>();
    private final Set<String> inclusions = new HashSet<>();
    private BiFunction<String, Object, Object> transformation;
    private boolean preserveStructure = true;
    private boolean flatten = false;
    private String flattenPrefix = "";

    BulkMappingBuilder(TargetBuilder parent, String sourcePattern) {
        this.parent = parent;
        this.sourcePattern = sourcePattern;
    }

    /**
     * Exclude specific paths from the bulk mapping.
     */
    public BulkMappingBuilder excluding(String... paths) {
        Collections.addAll(exclusions, paths);
        return this;
    }

    /**
     * Include only specific paths from the bulk mapping.
     */
    public BulkMappingBuilder including(String... paths) {
        Collections.addAll(inclusions, paths);
        return this;
    }

    /**
     * Apply a transformation to each match.
     */
    public BulkMappingBuilder transformEach(BiFunction<String, Object, Object> transformation) {
        this.transformation = transformation;
        return this;
    }

    /**
     * Apply a simple value transformation to each match.
     */
    public BulkMappingBuilder transformValues(Function<Object, Object> valueTransformation) {
        this.transformation = (key, value) -> valueTransformation.apply(value);
        return this;
    }

    /**
     * Flatten the structure instead of preserving hierarchy.
     */
    public BulkMappingBuilder flatten() {
        this.preserveStructure = false;
        this.flatten = true;
        return this;
    }

    /**
     * Flatten with a prefix for field names.
     */
    public BulkMappingBuilder flattenWithPrefix(String prefix) {
        this.preserveStructure = false;
        this.flatten = true;
        this.flattenPrefix = prefix;
        return this;
    }

    /**
     * Preserve the nested structure when mapping.
     */
    public BulkMappingBuilder preserveStructure() {
        this.preserveStructure = true;
        this.flatten = false;
        return this;
    }

    /**
     * Specify the target path.
     */
    public BulkTargetPathBuilder to(String targetPath) {
        return new BulkTargetPathBuilder(parent, sourcePattern, targetPath,
                exclusions, inclusions, transformation,
                preserveStructure, flatten, flattenPrefix);
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/bulk/BulkMappingBuilder.kt
package xyz.mahmoudahmed.dsl.bulk

import xyz.mahmoudahmed.dsl.builders.TargetBuilder

/**
 * Builder for bulk mapping operations with improved type safety.
 */
class BulkMappingBuilder(
    private val parent: TargetBuilder,
    private val sourcePattern: String,
    private val config: BulkMappingConfig = BulkMappingConfig()
) {
    /**
     * Exclude specific paths from the bulk mapping.
     */
    fun excluding(vararg paths: String): BulkMappingBuilder {
        require(paths.isNotEmpty()) { "At least one exclusion path must be provided" }
        val newConfig = config.copy(exclusions = config.exclusions + paths.toSet())
        return BulkMappingBuilder(parent, sourcePattern, newConfig)
    }

    /**
     * Include only specific paths from the bulk mapping.
     */
    fun including(vararg paths: String): BulkMappingBuilder {
        require(paths.isNotEmpty()) { "At least one inclusion path must be provided" }
        val newConfig = config.copy(inclusions = config.inclusions + paths.toSet())
        return BulkMappingBuilder(parent, sourcePattern, newConfig)
    }

    /**
     * Apply a transformation to each match with path context.
     */
    fun <T : Any, R : Any> transformEach(transformation: (path: String, value: T) -> R): BulkMappingBuilder {
        @Suppress("UNCHECKED_CAST")
        val typedTransformation = transformation as (String, Any) -> Any
        val newConfig = config.copy(transformation = typedTransformation)
        return BulkMappingBuilder(parent, sourcePattern, newConfig)
    }

    /**
     * Apply a simple value transformation to each match.
     */
    fun <T : Any, R : Any> transformValues(valueTransformation: (T) -> R): BulkMappingBuilder {
        @Suppress("UNCHECKED_CAST")
        val wrappedTransformation: (String, Any) -> Any = { _, value ->
            valueTransformation(value as T) as Any
        }
        val newConfig = config.copy(transformation = wrappedTransformation)
        return BulkMappingBuilder(parent, sourcePattern, newConfig)
    }

    /**
     * Flatten the structure instead of preserving hierarchy.
     */
    fun flatten(): BulkMappingBuilder {
        val newConfig = config.copy(preserveStructure = false, flatten = true)
        return BulkMappingBuilder(parent, sourcePattern, newConfig)
    }

    /**
     * Flatten with a prefix for field names.
     */
    fun flattenWithPrefix(prefix: String): BulkMappingBuilder {
        require(prefix.isNotBlank()) { "Flatten prefix cannot be blank" }
        val newConfig = config.copy(
            preserveStructure = false,
            flatten = true,
            flattenPrefix = prefix
        )
        return BulkMappingBuilder(parent, sourcePattern, newConfig)
    }

    /**
     * Preserve the nested structure when mapping.
     */
    fun preserveStructure(): BulkMappingBuilder {
        val newConfig = config.copy(preserveStructure = true, flatten = false)
        return BulkMappingBuilder(parent, sourcePattern, newConfig)
    }

    /**
     * Specify the target path.
     */
    fun to(targetPath: String): BulkTargetPathBuilder {
        require(targetPath.isNotBlank()) { "Target path cannot be blank" }
        return BulkTargetPathBuilder(parent, sourcePattern, targetPath, config)
    }
}